Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menu bar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menu bar, select Cell$\rightarrow$Run All).

Make sure that in addition to the code, you provide written answers for all questions of the assignment. 

Below, please fill in your name and collaborators:

In [1]:
NAME = "James Shen"


## Assignment 3 - Data Analysis using Pandas
**(15 points total)**

For this assignment, we will analyze the open dataset with data on the passengers aboard the Titanic.

The data file for this assignment can be downloaded from Kaggle website: https://www.kaggle.com/c/titanic/data, file `train.csv`. The definition of all variables can be found on the same Kaggle page, in the Data Dictionary section.

Read the data from the file into pandas DataFrame. Analyze, clean and transform the data to answer the following question: 

**What categories of passengers were most likely to survive the Titanic disaster?**

**Question 1.**  _(20 points)_
* The answer to the main question - What categories of passengers were most likely to survive the Titanic disaster? Conduct a detailed analysis to identify which passengers were most likely to survive. _(15 points)_
* The detailed explanation of the logic of the analysis _(5 points)_

**Question 2.**  _(5 points)_
* What attributes did you use for the analysis? Explain how you used them and why you decided to use them. 
* Provide a complete list of all attributes used.

**Question 3.**  _(5 points)_
* Were there any missing values? How did they impact your analysis? Consider dropping the rows with missing values using the drop() function. Did your findings change?


### Question 1a. 
#### What categories of passengers were most likely to survive the Titanic disaster? Conduct a detailed analysis to identify which passengers were most likely to survive

##### <font color='magenta'>First, load the dataset from the file 'train.csv' into the local computer memory.

In [2]:
import pandas as pd
df = pd.read_csv('train.csv')

##### <font color='magenta'>Then, take a look at the big picture of this dataset, including a summary of its statistics.

In [3]:
df.describe(include='all')

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000      204   
unique         NaN         NaN         NaN     681         NaN      147   
top            NaN         NaN         NaN  347082         NaN  B96 B98   
freq           NaN         NaN         NaN       7         NaN        4   
mean     29.699118    0.523008    0.381594     NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

##### <font color='magenta'>Next, take a look at the summary of its information, which includes a list of all column name, number of non-null values in each column, and data type of each column

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


##### <font color='magenta'>Also take a look at a few rows of the content of this dataset.

In [5]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

##### <font color='magenta'>In order to know whether  passenger 'Age' relates to the survival rate, I check its statistics for each survived group.

In [6]:
df.groupby(['Survived'])['Age'].agg(['min','max','mean','median'])

min   max       mean  median
Survived                               
0         1.00  74.0  30.626179    28.0
1         0.42  80.0  28.343690    28.0

##### <font color='magenta'>From the 'Age' variable statistics above, I could not find any salience for different survived group.

##### <font color='magenta'>Check the count of passengers class for each survived group to see if passengers class relates to the survival rate.

In [7]:
Survived_group = df.groupby(['Survived','Pclass'])['PassengerId'].count()
Survived_group

Survived  Pclass
0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
Name: PassengerId, dtype: int64

##### <font color='magenta'>Look like the 1st class passengers have more survival rate, now I compute the results of survival rate for all passenger class

In [8]:
All_group = df.groupby(['Pclass'])['PassengerId'].count()
class1_survival_rate = 100*Survived_group.loc[1][1] / All_group.loc[1]
class2_survival_rate = 100*Survived_group.loc[1][2] / All_group.loc[2]
class3_survival_rate = 100*Survived_group.loc[1][3] / All_group.loc[3]

print(f'The 1st class passengers\' survival rate: {class1_survival_rate:.2f}%')
print(f'The 2nd class passengers\' survival rate: {class2_survival_rate:.2f}%')
print(f'The 3rd class passengers\' survival rate: {class3_survival_rate:.2f}%')

The 1st class passengers' survival rate: 62.96%
The 2nd class passengers' survival rate: 47.28%
The 3rd class passengers' survival rate: 24.24%


##### <font color='magenta'>In order to know whether  passenger 'Fare' relates to the survival rate, I check its statistics for each survived group.

In [9]:
df.groupby(['Survived'])['Fare'].agg(['min','max','mean','median'])

min       max       mean  median
Survived                                  
0         0.0  263.0000  22.117887    10.5
1         0.0  512.3292  48.395408    26.0

##### <font color='magenta'>Here's how passenger 'Fare' impacts the survival rate:

Passengers who paid higher fare had higher survival rate. This also matches the above results, i.e. the 1st class passengers had higher survival rate then the 2nd class passengers, and the 2nd class passengers had higher survival rate then the 3rd class passengers.

##### <font color='magenta'>Check the count of passengers sex for each survived group to see if passengers sex relates to the survival rate.

In [10]:
Survived_group = df.groupby(['Survived','Sex'])['PassengerId'].count()
Survived_group

Survived  Sex   
0         female     81
          male      468
1         female    233
          male      109
Name: PassengerId, dtype: int64

##### <font color='magenta'>Look like the female passengers have more survival rate, now I compute the results of survival rate for both passenger sex

In [11]:
All_group = df.groupby(['Sex'])['PassengerId'].count()
female_survival_rate = 100*Survived_group.loc[1]['female'] / All_group.loc['female']
male_survival_rate = 100*Survived_group.loc[1]['male'] / All_group.loc['male']

print(f'Female passengers\' survival rate: {female_survival_rate:.2f}%')
print(f'Male passengers\' survival rate: {male_survival_rate:.2f}%')

Female passengers' survival rate: 74.20%
Male passengers' survival rate: 18.89%


##### <font color='magenta'>Conclusion of what categories of passengers were most likely to survive the Titanic disaster:

I found that female passengers were most likely to survive the Titanic disaster. The 1st class passengers had more chance to survive than the 2nd class passengers, and the 2nd class passengers had more chance to survive the 3rd class ones. In addition, the passengers who paid more fare had more chance to survive the Titanic disaster.

### Question 1b. 
#### <font color='magenta'>The detailed explanation of the logic of the analysis

The analysis is to identify what categories of passengers were most likely to survive the Titanic disaster. Therefore, the logic is to go through all possible variables, group them by the 'Survived' variable, and find any distinction (e.g. higher counts, higher average/median, etc.). Some variable like Passenger name, Cabin number, Embarked port, etc. should be excluded from the analysis as they are obviously won't have any impact to the results. I tried passenger age, and hope to see younger passengers have more survival rate; however, the statistics results told me that the passenger age does not have impact either.

### Question 2. 
#### What attributes did you use for the analysis? Explain how you used them and why you decided to use them. Provide a complete list of all attributes used.

<font color='magenta'>Below is a complete list of all attributes that I used for the analysis:

1. PassengerId: I used it in multiple places. The reason that I use it because it can be treated like the primary key which can uniquly identify each row of record. For example, when I count 'Survived' for each class, I can do so by the statements below:

> df.groupby(['Survived','Pclass'])['PassengerId'].count()
> df.groupby(['Survived','Sex'])['PassengerId'].count()


2. Survived: This is the key attribute to identify passengers who survived the Titanic disaster. The statements above shows how I used this attribute.

3. Pclass: This attributes represents ticket class (1 = 1st, 2 = 2nd, 3 = 3rd). I decided to use it because it could impact survival rate, and it turned out True. The statement above (the top one) shows how I used this attribute.

4. Sex: This attributes represents passenger sex (male or female). I decided to use it because it could impact survival rate, and it turned out True. The statement above (the bottom one) shows how I used this attribute.

5. Fare: This attributes represents passenger fare. It is a numerical value, which should be a kind of currency. It could be US dollar or UK pound or some other currencies. I decided to use it because it could impact survival rate, and it turned out True. The statements below shows how I used this attribute.

> df.groupby(['Survived'])['Fare'].agg(['min','max','mean','median'])

6. Age: This attributes represents passenger age in years. I decided to use it because it could impact survival rate, but it turned out False. The statements below shows how I used this attribute.

> df.groupby(['Survived'])['Age'].agg(['min','max','mean','median'])


### Question 3a. 
#### Were there any missing values? How did they impact your analysis?

In [12]:
df.isna().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

<font color='magenta'>Yes. There were missing values in the following columns:
* Age: Passenger age in years
* Cabin: Cabin number
* Embarked: Port of Embarkation

They can also be identified by the non-null value count. Most columns has 891 non-null elements, except the following columns:
* Age: 714
* Cabin: 204
* Embarked: 889

<font color='magenta'>The missing values in the Age column could impact my analysis, because I need them to compute Age statistics, but I did not find any salience.

<font color='magenta'>For Cabin, it's just a room number, so it does not impact my analysis. For Embarked, it only tells me where the passenger came on board, so it does not impact my analysis either.

### Question 3b. 
#### Consider dropping the rows with missing values using the drop() function. Did your findings change?

Since only the missing values in the 'Age' column might impact my analysis, I am going to drop rows with missing values in 'Age' column to see if there is any change.

##### <font color='magenta'>First, identify rows with missing values in 'Age' column:

In [13]:
missing_age = df.loc[df['Age'].isna()]
missing_age.head()

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S  
19      0      0    2649   7.2250   NaN        C  
26      0      0    2631   7.2250   NaN        C  
28      0      0  330959   7.8792   NaN        Q

##### <font color='magenta'>Before dropping rows, take a notes of the remaining rows that we expect to have:

In [14]:
df['PassengerId'].count() - missing_age['PassengerId'].count()

714

##### <font color='magenta'>Now we can use drop() function to drop those rows:

In [15]:
remaining_age = df.drop(index=df[df['Age'].isna()].index)
missing_age.head()

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S  
19      0      0    2649   7.2250   NaN        C  
26      0      0    2631   7.2250   NaN        C  
28      0      0  330959   7.8792   NaN        Q

##### <font color='magenta'>Check the count of the remaining rows that we expect to have:

In [16]:
remaining_age['PassengerId'].count()

714

##### <font color='magenta'>The non-null values in 'Age' column should have the same count as above:

In [17]:
remaining_age['Age'].count()

714

##### <font color='magenta'>Now, we can use this new DataFrame 'remaining_age' to check its statistics for each survived group again:

In [18]:
remaining_age.groupby(['Survived'])['Age'].agg(['min','max','mean','median'])

min   max       mean  median
Survived                               
0         1.00  74.0  30.626179    28.0
1         0.42  80.0  28.343690    28.0

##### <font color='magenta'>The above result told me that there is not any change at all after I dropped the missing values for the 'Age' column.

##### Below I would like to try dropna() function and see I can get the same result as remaining_age:

In [19]:
new_age = df[['PassengerId','Survived','Age']]
new_age

PassengerId  Survived   Age
0              1         0  22.0
1              2         1  38.0
2              3         1  26.0
3              4         1  35.0
4              5         0  35.0
..           ...       ...   ...
886          887         0  27.0
887          888         1  19.0
888          889         0   NaN
889          890         1  26.0
890          891         0  32.0

[891 rows x 3 columns]

##### Since there is no missing data for 'PassengerId' and 'Survived', the new_age only dropped rows with missing value in 'Age' column:

In [20]:
new_age = new_age.dropna()
new_age

PassengerId  Survived   Age
0              1         0  22.0
1              2         1  38.0
2              3         1  26.0
3              4         1  35.0
4              5         0  35.0
..           ...       ...   ...
885          886         0  39.0
886          887         0  27.0
887          888         1  19.0
889          890         1  26.0
890          891         0  32.0

[714 rows x 3 columns]

##### Compare DataFrame 'new_age' with the 3 columns in DataFrame 'remaining_age':

In [21]:
(new_age == remaining_age[['PassengerId','Survived','Age']]).all()

PassengerId    True
Survived       True
Age            True
dtype: bool

##### Above result tole me that they are identical

In [22]:
new_age.groupby(['Survived'])['Age'].agg(['min','max','mean','median'])

min   max       mean  median
Survived                               
0         1.00  74.0  30.626179    28.0
1         0.42  80.0  28.343690    28.0

##### The above result told me that there is not any change at all after I dropped the missing values for the 'Age' column.